scylla/src/transport/connection_pool.rs

```diff
@@ -245,6 +264,83 @@ impl NodeConnectionPool {
     }
 }
 
+struct Keepaliver {
+    connections: Arc<ArcSwap<MaybePoolConnections>>,
+    node_address: IpAddr, // This address is only used to enrich the log messages
+    keepalive_interval: Duration,
+}
+
+impl Keepaliver {
+    pub fn load_connections(&self) -> Vec<Arc<Connection>> {
+        use MaybePoolConnections::*;
+        use PoolConnections::*;
+
+        let pool = self.connections.load_full();
+        match &*pool {
+            Ready(NotSharded(conns)) => conns.clone(),
+            Ready(Sharded { connections, .. }) => connections.iter().flatten().cloned().collect(),
+            Initializing => vec![],
+            Broken => {
+                debug!("Cannot send connection keepalives for node {} as there are no alive connections in the pool", self.node_address);
+                vec![]
+            }
+        }
+    }
+
+    async fn work(self) {
+        let mut interval = tokio::time::interval(self.keepalive_interval);
+        interval.tick().await; // Use up the first, instant tick.
+
+        // Default behaviour (Burst) is not suitable for sending keepalives.
+        use tokio::time::MissedTickBehavior::Delay;
+        interval.set_missed_tick_behavior(Delay);
+
+        // Wait for the second tick (so that `self.keepalive_interval` time passes since entering
+        // this function)
+        interval.tick().await;
+
+        loop {
+            let send_keepalives = self.send_keepalives();
+
+            tokio::select! {
+                _ = send_keepalives => {
+                    // Sending keepalives finished before receiving new tick.
+                    // Wait for the new tick and start over.
+                    interval.tick().await;
+                }
+                _ = interval.tick() => {
+                    // New tick arrived before `send_keepalives` was finished.
+                    // Stop polling `send_keepalives` and start over.
+                    //
+                    // `Interval::tick()` is cancellation safe, so it's ok to use it like that.
+                }
+            }
+        }
+    }
+
+    async fn send_keepalives(&self) {
+        let connections = self.load_connections();
+        let mut futures = connections
+            .into_iter()
+            .map(Self::send_keepalive_query)
+            .collect::<FuturesUnordered<_>>();
+
+        while futures.next().await.is_some() {}
```

Comment:
AFAIK, here we will wait for the first failure and then stop waiting for remaining futures. Maybe we should wait until all of them complete so that we won't miss any errors (they are printed in logs)? WDYT @psarna ?

Reply:
yup, sounds better

Reply:
I don't think that this code waits for the first failure. It is impossible to say whether an execution of `Self::send_keepalive_query` failed - this function does not return anything.

```rust
while futures.next().await.is_some() {}
```
`.next()` comes from `Stream` - It creates a future that resolves to the next item in the stream.
If there are no items left in the stream, created future will resolve to `None`.

So from what I know, the code above resolves all previously created futures.

Reply:
Yes, you are right.